Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp

```diff
@@ -1380,296 +1471,361 @@ void ADefiWalletCoreActor::Erc20Allowance(FString contractAddress,
+            std::string myfromaddress = TCHAR_TO_UTF8(*fromAddress);
+            std::string mytoaddress = TCHAR_TO_UTF8(*toAddress);
+            std::string mycronosrpc = TCHAR_TO_UTF8(*myCronosRpc);
+            std::string mytokenid = TCHAR_TO_UTF8(*tokenid);
+
+            ::rust::Vec<::std::uint8_t> myadditionaldata;
+            myadditionaldata.clear();
+            for (int i = 0; i < additionaldata.Num(); i++) {
+              myadditionaldata.push_back(additionaldata[i]);
+            }
+            assert(myadditionaldata.size() == additionaldata.Num());
+
+            Erc721 erc721 =
+                new_erc721(mycontractaddress, mycronosrpc, myCronosChainID)
+                    .legacy();
+            ::org::defi_wallet_core::CronosTransactionReceiptRaw receipt =
+                erc721.safe_transfer_from_with_data(myfromaddress, mytoaddress,
+                                                    mytokenid, myadditionaldata,
+                                                    *privatekey);
+            convertCronosTXReceipt(receipt, txresult);
+          }
+
+        } catch (const rust::cxxbridge1::Error &e) {
+          result = FString::Printf(
+              TEXT("CronosPlayUnreal Erc721SafeTransferFromWithData Error: %s"),
+              UTF8_TO_TCHAR(e.what()));
+        }
+
+        AsyncTask(ENamedThreads::GameThread, [Out, txresult, result]() {
+          Out.ExecuteIfBound(txresult, result);
+        });
+      });
 }
 
-void ADefiWalletCoreActor::Erc721Approve(
-    FString contractAddress, int walletindex, FString approvedAddress,
-    FString tokenid, FCronosTransactionReceiptRaw &result, bool &success,
-    FString &output_message) {
-  try {
-    if (NULL == _coreWallet) {
-      success = false;
-      output_message = TEXT("Invalid Wallet");
-      return;
-    }
-
-    char hdpath[100];
-    snprintf(hdpath, sizeof(hdpath), "m/44'/%d'/0'/0/%d", EthCoinType,
-             walletindex);
-    rust::cxxbridge1::Box<PrivateKey> privatekey = _coreWallet->get_key(hdpath);
-
-    std::string mycontractaddress = TCHAR_TO_UTF8(*contractAddress);
-    std::string myapprovedAddress = TCHAR_TO_UTF8(*approvedAddress);
-    std::string mycronosrpc = TCHAR_TO_UTF8(*myCronosRpc);
-    std::string mytokenid = TCHAR_TO_UTF8(*tokenid);
-
-    Erc721 erc721 =
-        new_erc721(mycontractaddress, mycronosrpc, myCronosChainID).legacy();
-    ::org::defi_wallet_core::CronosTransactionReceiptRaw receipt =
-        erc721.approve(myapprovedAddress, mytokenid, *privatekey);
-    convertCronosTXReceipt(receipt, result);
-    success = true;
-
-  } catch (const rust::cxxbridge1::Error &e) {
-    success = false;
-    output_message =
-        FString::Printf(TEXT("CronosPlayUnreal Erc721Approve Error: %s"),
-                        UTF8_TO_TCHAR(e.what()));
-  }
+void ADefiWalletCoreActor::Erc721Approve(FString contractAddress,
+                                         int walletindex,
```

Comment:
some functions had `int32`, others `int` for wallet index... is there a reason for it?

Reply:
because int length varies by platform, int32 seems more accurate
ref: https://docs.unrealengine.com/4.27/en-US/ProductionPipelines/DevelopmentSetup/CodingStandard/

Reply:
will change to int32